src/Common/Crypto/X509Certificate.h

```diff
@@ -0,0 +1,333 @@
+    }
+
+    std::string serialNumber() const
+    {
+        ASN1_INTEGER * serial = X509_get_serialNumber(certificate);
+        BIGNUM * bn = ASN1_INTEGER_to_BN(serial, nullptr);
+        char * hex = BN_bn2hex(bn);
+        std::string result(hex);
+        OPENSSL_free(hex);
+        BN_free(bn);
+        return result;
+    }
+
+    std::string signatureAlgorithm() const
+    {
+        const X509_ALGOR * sig_alg = X509_get0_tbs_sigalg(certificate);
+        char buffer[NAME_BUFFER_SIZE];
+        OBJ_obj2txt(buffer, sizeof(buffer), sig_alg->algorithm, 0);
+        return buffer;
+    }
+
+    std::string issuerName() const
+    {
+        char buffer[NAME_BUFFER_SIZE];
+        X509_NAME_oneline(X509_get_issuer_name(certificate), buffer, sizeof(buffer));
+        return buffer;
+    }
+
+    std::string subjectName() const
+    {
+        char buffer[NAME_BUFFER_SIZE];
+        X509_NAME_oneline(X509_get_subject_name(certificate), buffer, sizeof(buffer));
+        return buffer;
+    }
+
+    std::string issuerName(uint nid) const
+    {
+        if (X509_NAME * issuer = X509_get_issuer_name(certificate))
+        {
+            char buffer[NAME_BUFFER_SIZE];
+            if (X509_NAME_get_text_by_NID(issuer, nid, buffer, sizeof(buffer)) >= 0)
+                return std::string(buffer);
+        }
+        return std::string();
+    }
+
+    std::string subjectName(uint nid) const
+    {
+        if (X509_NAME * subj = X509_get_subject_name(certificate))
+        {
+            char buffer[NAME_BUFFER_SIZE];
+            if (X509_NAME_get_text_by_NID(subj, nid, buffer, sizeof(buffer)) >= 0)
+                return std::string(buffer);
+        }
+        return std::string();
+    }
+
+    std::string commonName() const { return subjectName(NID_commonName); }
+
+    std::string publicKeyAlgorithm() const
+    {
+        EVP_PKEY * pkey = X509_get_pubkey(certificate);
+        if (!pkey)
+            return {};
+
+        int nid = EVP_PKEY_id(pkey);
+        EVP_PKEY_free(pkey);
+
+        if (nid == NID_undef)
+            return {};
+
+        char buf[128] = {0};
+        OBJ_obj2txt(buf, sizeof(buf), OBJ_nid2obj(nid), 0);
+        return std::string(buf);
+    }
+
+    std::string validFrom() const
+    {
+        ASN1_TIME * valid_from = X509_get_notBefore(certificate);
+        return reinterpret_cast<char *>(valid_from->data);
+    }
+
+    std::string expiresOn() const
+    {
+        ASN1_TIME * not_before = X509_get_notBefore(certificate);
+        return reinterpret_cast<char *>(not_before->data);
+    }
+
+    class Subjects
+    {
+    public:
+        using container = boost::container::flat_set<String>;
+        enum class Type
+        {
+            CN,
+            SAN
+        };
+
+    private:
+        std::array<container, size_t(Type::SAN) + 1> subjects;
+
+    public:
+        inline const container & at(Type type_) const { return subjects[static_cast<size_t>(type_)]; }
```

Comment:
(Minor: `inline` is implicit for in-class functions. Could omit it. Moving everything into source file would be even better :-) )